mixnet/node.py

```diff
@@ -1,107 +1,159 @@
+        self.global_config = global_config
+        self.mixgossip_channel = GossipChannel(config.gossip, self.__process_msg)
+        self.reconstructor = MessageReconstructor()
+        self.broadcast_channel = asyncio.Queue()
 
-    async def __run(self):
-        """
-        Read SphinxPackets from inbound socket and spawn a thread for each packet to process it.
+        sample_packet, _ = PacketBuilder.build_real_packets(
+            bytes(1), global_config.membership, self.global_config.max_mix_path_length
+        )[0]
+        self.packet_size = len(sample_packet.bytes())
 
-        This thread approximates a M/M/inf queue.
+    async def __process_msg(self, msg: bytes) -> bytes | None:
+        """
+        A handler to process messages received via gossip channel
+        """
+        flag, msg = Node.__parse_msg(msg)
+        match flag:
+            case MsgType.NOISE:
+                # Drop noise packet
+                return None
+            case MsgType.REAL:
+                # Handle the packet and gossip the result if needed.
+                sphinx_packet = SphinxPacket.from_bytes(msg)
+                new_sphinx_packet = await self.__process_sphinx_packet(sphinx_packet)
+                if new_sphinx_packet is None:
+                    return None
+                return Node.__build_msg(MsgType.REAL, new_sphinx_packet.bytes())
+
+    async def __process_sphinx_packet(
+        self, packet: SphinxPacket
+    ) -> SphinxPacket | None:
+        """
+        Unwrap the Sphinx packet and process the next Sphinx packet or the payload.
+        """
+        try:
+            processed = packet.process(self.config.private_key)
+            match processed:
+                case ProcessedForwardHopPacket():
+                    return processed.next_packet
+                case ProcessedFinalHopPacket():
+                    await self.__process_sphinx_payload(processed.payload)
+        except ValueError:
+            # Return SphinxPacket as it is, if it cannot be unwrapped by the private key of this node.
+            return packet
+
+    async def __process_sphinx_payload(self, payload: Payload):
+        """
+        Process the Sphinx payload and broadcast it if it is a real message.
+        """
+        msg_with_flag = self.reconstructor.add(
+            Fragment.from_bytes(payload.recover_plain_playload())
+        )
+        if msg_with_flag is not None:
+            flag, msg = PacketBuilder.parse_msg_and_flag(msg_with_flag)
+            if flag == MessageFlag.MESSAGE_FLAG_REAL:
+                await self.broadcast_channel.put(msg)
+
+    def connect(self, peer: Node):
         """
+        Establish a duplex connection with a peer node.
+        """
+        noise_msg = Node.__build_msg(MsgType.NOISE, bytes(self.packet_size))
+        inbound_conn, outbound_conn = asyncio.Queue(), asyncio.Queue()
+
+        # Register a duplex connection for its own use
+        self.mixgossip_channel.add_conn(
```

Comment:
We can use "nomssip" name for out variant of gossip protocol ;)

Reply:
Good naming. 953b2d6 renamed it.